include/swift/SIL/SILDifferentiabilityWitness.h

```diff
@@ -0,0 +1,132 @@
+//===--- SILDifferentiabilityWitness.h - Differentiability witnesses ------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the SILDifferentiabilityWitness class, which maps an
+// original SILFunction and derivative configuration (parameter indices, result
+// indices, derivative generic signature) to derivative functions (JVP and VJP).
+//
+// SIL differentiability witnesses are generated from the `@differentiable`
+// and `@differentiating` attributes AST declaration attributes.
+// Differentiability witnesses are canonicalized by the differentiation SIL
+// transform, which fills in missing derivative functions. Canonical
+// differentiability witnesses from other modules can be deserialized to look up
+// derivative functions.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SIL_SILDIFFERENTIABILITYWITNESS_H
+#define SWIFT_SIL_SILDIFFERENTIABILITYWITNESS_H
+
+#include "swift/AST/AutoDiff.h"
+#include "swift/AST/GenericSignature.h"
+#include "swift/SIL/SILAllocated.h"
+#include "swift/SIL/SILInstruction.h"
+#include "llvm/ADT/ilist_node.h"
+#include "llvm/ADT/ilist.h"
+
+namespace swift {
+
+class SILPrintContext;
+
+class SILDifferentiabilityWitness
+    : public llvm::ilist_node<SILDifferentiabilityWitness>,
+      public SILAllocated<SILDifferentiabilityWitness>
+{
+private:
+  /// The module which contains the differentiability witness.
```

Comment:
You also need to store a source location here for differentiation diagnostics.

Reply:
Chatted in person, storing source location may not make the most sense.
Added `DeclAttribute *` field (representing AST `@differentiable` or `@differentiating attribute) in https://github.com/apple/swift/pull/27487/commits/4ddcd3f059b686dcc6137bf4a7455419b94798aa.